src/memory-analyzer/memory_analyzer_parse_options.cpp

```diff
@@ -1,102 +1,188 @@
-// Copyright 2018 Author: Malte Mues <mail.mues@gmail.com>
+/*******************************************************************\
+
+Module: Memory Analyzer
+
+Author: Malte Mues <mail.mues@gmail.com>
+        Daniel Poetzl
+
+\*******************************************************************/
 
 /// \file
 /// Commandline parser for the memory analyzer executing main work.
 
-#ifdef __linux__
+// clang-format off
+#if defined(__linux__) || \
+    defined(__FreeBSD_kernel__) || \
+    defined(__GNU__) || \
+    defined(__unix__) || \
+    defined(__CYGWIN__) || \
+    defined(__MACH__)
+// clang-format on
 
 #include "memory_analyzer_parse_options.h"
 #include "analyze_symbol.h"
 #include "gdb_api.h"
 
+#include <fstream>
+
+#include <ansi-c/ansi_c_language.h>
+
 #include <goto-programs/goto_model.h>
 #include <goto-programs/read_goto_binary.h>
+#include <goto-programs/show_symbol_table.h>
+
+#include <langapi/mode.h>
+
 #include <util/config.h>
 #include <util/exit_codes.h>
 #include <util/message.h>
 #include <util/string_utils.h>
+#include <util/version.h>
+
+memory_analyzer_parse_optionst::memory_analyzer_parse_optionst(
+  int argc,
+  const char *argv[])
+  : parse_options_baset(
+      MEMORY_ANALYZER_OPTIONS,
+      argc,
+      argv,
+      ui_message_handler),
+    ui_message_handler(cmdline, std::string("MEMORY-ANALYZER ") + CBMC_VERSION),
+    message(ui_message_handler)
+{
+}
 
 int memory_analyzer_parse_optionst::doit()
 {
-  // This script is the entry point and has to make sure
-  // that the config object is set to a valid architecture.
-  // config is later used to determine right size for types.
-  // If config is not set, you might see a bunch of types with
-  // size 0.
-  // It might be desierable to convert this into flags in the future.
-  config.ansi_c.mode = configt::ansi_ct::flavourt::GCC;
-  config.ansi_c.set_arch_spec_x86_64();
+  if(cmdline.isset("version"))
+  {
+    message.status() << CBMC_VERSION << '\n';
+    return CPROVER_EXIT_SUCCESS;
+  }
+
+  config.set(cmdline);
 
   if(cmdline.args.size() != 1)
   {
-    error() << "Please provide one binary with symbols to process" << eom;
-    return CPROVER_EXIT_USAGE_ERROR;
+    throw invalid_command_line_argument_exceptiont(
+      "no binary provided for analysis", "<binary>");
+  }
+
+  if(!cmdline.isset("symbols"))
+  {
+    throw invalid_command_line_argument_exceptiont(
+      "need to provide symbols to analyse via --symbols", "--symbols");
+  }
+
+  const bool core_file = cmdline.isset("core-file");
+  const bool breakpoint = cmdline.isset("breakpoint");
+
+  if(!(core_file ^ breakpoint))
+  {
+    throw invalid_command_line_argument_exceptiont(
+      "need to provide either option --core-file or option --breakpoint", "");
```

Comment:
I think it is slightly unclear from the message that the `--core-file` and `--breakpoint` options are mutually exclusive - maybe two separate error messages for when neither or both are set?

Reply:
Ok.